### PR TITLE
Trading Pair Min volume

### PR DESCRIPTION
**Pull request by nakul1010**
The Pr fixes following things
- it doesn't allow a user to register pair which has volume less than min volume. 
- Min Volume = Min price * Min qty.
- The min price, min qty, min volume a pair can be register is with 10000.

### CODE REVIEW

File: pallets/ocex/src/lib.rs [hunk 1]
```diff
@@ -366,6 +366,17 @@ pub mod pallet {
...
 			);
+			//enclave will only support min volume of 10^-8, if the trading pair register
+			//falls below min volume limits it will send out an error
+			ensure!(
+				(Decimal::from(min_order_price.saturated_into::<u128>())
+					.div(&Decimal::from(UNIT_BALANCE)))
+				.saturating_mul(
+					Decimal::from(min_order_qty.saturated_into::<u128>())
+						.div(&Decimal::from(UNIT_BALANCE))
+				) > Decimal::new(1, 8),
```
Comment (serhii-temchenko) [inline pallets/ocex/src/lib.rs]: If `Decimal::new(1, 8)` is a minimum value - please replace with `>=`

File: pallets/ocex/src/lib.rs [hunk 2]
```diff
@@ -366,6 +366,17 @@ pub mod pallet {
...
+				Error::<T>::StorageOverflow,
```
Comment (serhii-temchenko) [inline pallets/ocex/src/lib.rs]: Is't it `underflow`?

**Review by serhii-temchenko — CHANGES_REQUESTED** [review]

File: pallets/ocex/src/lib.rs (hunk 2 above)
Comment (nakul1010) [reply pallets/ocex/src/lib.rs]: Added InvalidTradingPairConfig value in Enum

File: pallets/ocex/src/lib.rs [hunk 3]
```diff
@@ -366,6 +368,17 @@ pub mod pallet {
 				!<TradingPairs<T>>::contains_key(quote, base),
 				Error::<T>::TradingPairAlreadyRegistered
 			);
+			//enclave will only support min volume of 10^-8, if the trading pair register
+			//falls below min volume limits it will send out an error
+			ensure!(
+				(Decimal::from(min_order_price.saturated_into::<u128>())
+					.div(&Decimal::from(UNIT_BALANCE)))
```
Comment (Gauthamastro) [inline pallets/ocex/src/lib.rs]: Decimal::from will panic if min_order_price is greater than Decimal::MAX.

**Review by Gauthamastro — CHANGES_REQUESTED** [review]

File: pallets/ocex/src/lib.rs (hunk 3 above)
Comment (nakul1010) [reply pallets/ocex/src/lib.rs]: collect_fees, claim_withdraw, deposit, register_trading_pair all have Decimal::from should we convert it to Decimal::from_f64 and handle the error it same thing as we have done [here in OB](https://github.com/Polkadex-Substrate/orderbook/blob/72ff75be699eb56983ac488db45845e1e29439bc/primitives/src/orderbook.rs#L269)

Comment (felixfaisal) [reply pallets/ocex/src/lib.rs]: This is present in #522 
I've added a check that ensures `Decimal::from` will not panic, as we return an error if any param is greater than `10^27`.

Comment (nakul1010) [reply pallets/ocex/src/lib.rs]: Resolved checked for min value of 1000 u128

File: pallets/ocex/src/lib.rs [hunk 4]
```diff
@@ -408,6 +411,18 @@ pub mod pallet {
...
+			//enclave will only support min volume of 10^-8
+			//if trading pairs volume falls below it will pass a UnderFlow Error
+			ensure!(
+				min_order_price.saturated_into::<u128>() > MIN_CONFIG &&
+					min_order_qty.saturated_into::<u128>() > MIN_CONFIG &&
+					min_order_price
+						.saturated_into::<u128>()
+						.saturating_mul(min_order_qty.saturated_into::<u128>()) >
+						MIN_CONFIG,
+				Error::<T>::TradingPairConfigUnderflow
```
Comment (serhii-temchenko) [inline pallets/ocex/src/lib.rs]: A bit confused by this assertion.
1) First here you are checking that `min_order_price` is bigger than `MIN_CONFIG` 
2) Then you are checking that `min_order_qty` is bigger than `MIN_CONFIG`
3) And then you are checking that `min_order_price` * `min_order_qty` is bigger then `MIN_CONFIG`
Right?
But question - if you already asserted 1st and 2nd points - how 3rd can be false? By description `Min Volume = Min price * Min qty` but with this assertion - minimum volume always will be  `MIN_CONFIG` * `MIN_CONFIG` - other ways you are throwing  `TradingPairConfigUnderflow` error. Isn't just `min_order_price.saturated_into::<u128>().saturating_mul(min_order_qty.saturated_into::<u128>()) > MIN_CONFIG` is enough (if we need assert that resulting value of  `min_order_price` * `min_order_qty` should be bigger then `MIN_CONFIG`)?

File: pallets/ocex/src/lib.rs [hunk 5]
```diff
@@ -49,6 +49,7 @@ type BalanceOf<T> =
 	<<T as Config>::NativeCurrency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
 
 const DEPOSIT_MAX: u128 = 1_000_000_000_000_000_000_000_000_000;
+const MIN_CONFIG: u128 = 10000;
```
Comment (serhii-temchenko) [inline pallets/ocex/src/lib.rs]: Please change this name to more self describing one e.g. `TRADE_OPERATION_MIN_VALUE` or something...

File: pallets/ocex/src/tests.rs [hunk 6]
```diff
@@ -651,16 +634,36 @@ fn test_update_trading_pair_bad_origin() {
...
+				1_u128.into(),
+				100_u128.into(),
+				100_u128.into(),
+				10_u128.into(),
+			),
+			Error::<Test>::TradingPairConfigUnderflow
+		);
+	});
+}
+
```
Comment (serhii-temchenko) [inline pallets/ocex/src/tests.rs:673]: Please add tests where you are testing edges (e.g. `10001`, `999`). I can see that you are fixed other tests with `10001` but for this one should be defined dedicated test because in other cases requirements could change in advance and this case will be uncovered...

**Review by serhii-temchenko — CHANGES_REQUESTED** [review]

File: pallets/ocex/src/tests.rs (hunk 6 above)
Comment (serhii-temchenko) [reply pallets/ocex/src/tests.rs:673]: And these tests should fail or pass after `min_order_price * min_order_qty` operation so please hold this in mind..

File: pallets/ocex/src/lib.rs (hunk 4 above)
Comment (nakul1010) [reply pallets/ocex/src/lib.rs]: Yes you are right, if min_price and min_qty are checked the min_volume check would work by default. Should i remove it @Gauthamastro.

File: pallets/ocex/src/lib.rs (hunk 5 above)
Comment (nakul1010) [reply pallets/ocex/src/lib.rs]: Done

File: pallets/ocex/src/tests.rs (hunk 6 above)
Comment (nakul1010) [reply pallets/ocex/src/tests.rs:673]: Added test case test_register_trading_pair_volume_too_low for testing failing edge case,and  test_open_trading_pair already has the passing edge case

File: pallets/ocex/src/lib.rs (hunk 4 above)
Comment (Gauthamastro) [reply pallets/ocex/src/lib.rs]: We are not asserting a min value for min_price and min_qty, as it can be 1  ( u128 ) and MIN_CONFIG for min_qty. Then, that check will still pass, but our downstream libraries will break. We have to ensure at least as a sanity check that min_price, min_qty and min_volume doesn't fall under MIN_CONFIG.

Comment (nakul1010) [reply pallets/ocex/src/lib.rs]: Ok, so @serhii-temchenko we will keep it.

**Review by serhii-temchenko — DISMISSED** [review]

File: pallets/ocex/src/tests.rs [hunk 7]
```diff
@@ -1,20 +1,3 @@
-// This file is part of Polkadex.
-
-// Copyright (C) 2020-2022 Polkadex oü.
-// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
-
-// This program is free software: you can redistribute it and/or modify
-// it under the terms of the GNU General Public License as published by
-// the Free Software Foundation, either version 3 of the License, or
```
Comment (Gauthamastro) [inline pallets/ocex/src/tests.rs]: Why is this removed ?

**Review by Gauthamastro — CHANGES_REQUESTED** [review]

File: pallets/ocex/src/tests.rs (hunk 7 above)
Comment (nakul1010) [reply pallets/ocex/src/tests.rs]: the comment is added back

**Review by Gauthamastro — APPROVED** [review]
